src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+
+# (::VariationalObjective)(vi::VariationalInference, model::Model, num_samples, args...; kwargs...) = begin
+# end
+
+
+"""
+    grad!(vo, vi::VariationalInference, q::VariationalPosterior, model::Model, θ, out, args...)
+
+Computes the gradients used in `optimize!`. Default implementation is provided for `VariationalInference{AD}` where `AD` is either `ForwardDiffAD` or `TrackerAD`. This implicitly also gives a default implementation of `optimize!`.
+
+Variance reduction techniques, e.g. control variates, should be implemented in this function.
+"""
+grad!(vo, vi::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ, out, args...) where AD =
+    error("Turing.Variational.grad!: unmanaged variational inference algorithm: $(typeof(alg))")
+
+"""
+    vi(model::Model, alg::VariationalInference)
+    vi(model::Model, alg::VariationalInference, q::VariationalPosterior)
+
+Constructs the variational posterior from the `model` and performs the optimization following the configuration of the given `VariationalInference` instance.
+"""
+vi(model::Model, alg::VariationalInference) = 
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+vi(model::Model, alg::VariationalInference, q::VariationalPosterior) =
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+
+# default implementations
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: ForwardDiffAD}
+    # TODO: this probably slows down executation quite a bit; exists a better way of doing this?
+    f(θ_) = - vo(alg, q, model, θ_, args...)
+
+    chunk_size = getchunksize(alg)
+    # Set chunk size and do ForwardMode.
+    chunk = ForwardDiff.Chunk(min(length(θ), chunk_size))
+    config = ForwardDiff.GradientConfig(f, θ, chunk)
+    ForwardDiff.gradient!(out, f, θ, config)
+end
+
+# TODO: implement for `Tracker`
+# function grad(vo::ELBO, alg::ADVI, q::MeanField, model::Model, f, autodiff::Val{:backward})
+#     vo_tracked, vo_pullback = Tracker.forward()
+# end
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: TrackerAD}
+    θ_tracked = Tracker.param(θ)
+    y = - vo(alg, q, model, θ_tracked, args...)
+    Tracker.back!(y, 1.0)
+
+    DiffResults.value!(out, Tracker.data(y))
+    DiffResults.gradient!(out, Tracker.grad(θ_tracked))
+end
+
+"""
+    optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad())
+
+Iteratively updates parameters by calling `grad!` and using the given `optimizer` to compute the steps.
+"""
+function optimize!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ; optimizer = ADAGrad()) where AD
```

Comment:
Is there a particular reason that we're defaulting to `AdaGrad` here?

Reply:
Usually the one (+ `RMSProp`) used for `ADVI` in the literature from what I've seen. I remember just trying a couple of different optimizers to make sure the code worked.  The few others I tried didn't converge so I just defaulted to something I knew worked.

Reply:
Hmm interesting. Did you try Adam / AMSGrad?

Reply:
I remember trying ADAM and pretty sure it failed to converge. I believe it's because ADAM attempts to estimate second-order information and in ADVI you usually use quite few samples to estimate the ELBO. This then leads to high variance for the gradient estimate, in which case estimates second-order information can be quite unreliable.

Definitively something to explore further though.